src/clj/coffi/mem.clj

```diff
@@ -1350,3 +1766,523 @@
 (s/fdef defalias
   :args (s/cat :new-type qualified-keyword?
                :aliased-type any?))
+
+(defn- with-c-layout
+  "Forces a struct specification to C layout rules.
+
+  This will add padding fields between fields to match C alignment
+  requirements."
+  [struct-spec]
+  (let [aligned-fields
+        (loop [offset 0
+               aligned-fields []
+               fields (nth struct-spec 1)]
+          (if (seq fields)
+            (let [[[_ type :as field] & fields] fields
+                  size (size-of type)
+                  align (align-of type)
+                  r (rem offset align)]
+              (recur (cond-> (+ offset size)
+                       (pos? r) (+ (- align r)))
+                     (cond-> aligned-fields
+                       (pos? r) (conj [:coffi.layout/padding [:coffi.mem/padding (- align r)]])
+                       :always (conj field))
+                     fields))
+            (let [strongest-alignment (reduce max (map (comp align-of second) (nth struct-spec 1)))
+                  r (rem offset strongest-alignment)]
+              (cond-> aligned-fields
+                (pos? r) (conj [:coffi.layout/padding [:coffi.mem/padding (- strongest-alignment r)]])))))]
+    (assoc struct-spec 1 aligned-fields)))
+
+(defn- coffitype->typename [in]
+  (let [[arr _type n & {:keys [raw?] :as opts}] (if (vector? in) in [:- in])
+        arr? (= arr ::array)
+        array-types  {::byte   'bytes
+                      ::short  'shorts
+                      ::int    'ints
+                      ::long   'longs
+                      ::char   'chars
+                      ::float  'floats
+                      ::double 'doubles}
+        single-types {::byte     'byte
+                      ::short    'short
+                      ::int      'int
+                      ::long     'long
+                      ::char     'char
+                      ::float    'float
+                      ::double   'double
+                      ::c-string 'String}]
+    (cond (and arr? raw?) (get array-types _type 'objects)
+          (and arr?)      `clojure.lang.IPersistentVector
+          :default        (get single-types _type (keyword (str *ns*) (str _type))))))
```

Comment:
What does this default do? Calling `coffitype->typename` with something not in the specific lists here would result in something like `:some.ns/:some.ns/foo` which is unreadable.

Since this gets used to create tags for typehinting, I think that this only makes much sense if it's either the struct typename of another struct object, or one of these primitives. Should this do something to only produce a value in those cases?

Reply:
oh, i believe this is a remnant from when the defstruct macro didn't use a namespaced keyword internally, so this was supposed to add the namespace to the typename. since the input here is always a correctly namespaced keyword anyway (if not some primitive), this can get simplified to `(get single-types type type)`. it didn't cause any problems so i didn't notice it. fixed in [cb3e620](https://github.com/IGJoshua/coffi/pull/14/commits/cb3e62057ffa1606f4e66ef3df8d6654deaa295e)